text/3808-register-tool.md

```diff
@@ -0,0 +1,271 @@
+
+Please do *not* use tool attributes for metadata that changes the meaning of the code. At that point you are parsing a dialect of Rust, and there is no indication for your users that their code will be interpreted differently by your tool than by the compiler.  For example, `#[must_use]` and `#[automatically_derived]` would be suitable for tool attributes, but `#[repr]` and `#[panic_handler]` are not, because they change the meaning of the code. For that use case, use proc-macros, generated code, or bare (un-namespaced) attributes instead, all of which will give a hard error if they cannot be understood by the compiler. If absolutely necessary to use bare attributes, use a C-style namespace like `#[rustc_const_stable]`.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Language
+
+### Background: name resolution of preludes
+
+Currently, names in the type namespace are resolved in [the following order](https://github.com/rust-lang/reference/pull/1765):
+
+1. Explicit definitions (including imports)
+2. [Extern prelude] (crates injected using either `extern crate` or `--extern`)
+3. [Tool prelude][builtin-tools]
+4. Standard library prelude
+5. Language prelude
+
+[Extern prelude]: https://doc.rust-lang.org/nightly/reference/names/preludes.html#extern-prelude
+
+Note that currently, loading a crate completely prevents using a tool attribute with that name. Consider this program:
+
+```rust
+extern crate rustfmt; // or --extern rustfmt
+
+#[rustfmt::skip]  // ERROR could not find `skip` in `rustfmt`
+fn foo ( ) { }
+```
+
+### Semantics
+
+The tool prelude is separated into the tool attribute prelude (which is in the type namespace) and the lint prelude (which is only active inside lint controls).
+
+The `#![register_attribute_tool(ident)]` crate-level attribute adds a new tool to the tool attribute prelude. The `#![register_lint_tool(ident)]` crate-level attribute adds a new tool to the lint prelude. For both attributes, tools must be a single ident, not a nested path.
+
+`register_attribute_tool` and `register_lint_tool` are idempotent; duplicating the attribute has no effect.
+
+Crate-level macros such as `#![cfg_attr(...)]` may expand to `register_*_tool`, subject to all existing rules for macro expansion.
+
+Like today, attributes and lints in a tool namespace are always considered used by the compiler. The compiler does not verify the contents of any tool attribute, except to verify that all attributes are syntactically valid [tool attributes].
+
+Registering a predefined tool (`clippy`, `miri`, etc.) using `#![register_*_tool(...)]` is an error.
+
+The `rustc` tool namespace is currently reserved and will continue to be reserved after this RFC, i.e, `register_*_tool(rustc)` is an error.
+Additionally, any tool namespace starting with `rustc`, such as `rustc_lint`, is reserved.
```

Comment:
does this include names like `rustcat`

Reply:
yeah. i am not a fan of the current attribute gating but i don't want to touch it in this rfc.